// @framer/bredowmax.pagination/code/AreaPagination.tsx
import * as React from "react";
import { Frame, Page, PropertyControls, ControlType, PageEffect } from "framer";

export interface Props { width: number; height: number; ratio: number; backgroundColor: string; forwardColor: string; backwardColor: string; boolean: boolean; } // Define type of property

export class AreaPagination extends React.Component<Props> {
  constructor(props) {
    super(props);
    this.state = { currentPageIndex: 0, i: 0 };
    this.tapForwards = this.tapForwards.bind(this);
    this.tapBackwards = this.tapBackwards.bind(this);
  }

  static defaultProps = { ...Frame.defaultProps, width: 375, height: 812, ratio: 25, backgroundColor: "#000000", backwardColor: "hsla(0, 0%, 0%, 0.1)", forwardColor: "hsla(0, 0%, 100%, 0.1)", boolean: true}; // Set default properties

  static propertyControls: PropertyControls = { // Items shown in property panel
    ratio: { type: ControlType.Number, title: "Ratio", min: 5, max: 95 },
    backgroundColor: { type: ControlType.Color, title: "Background" },
    backwardColor: { type: ControlType.Color, title: "Backward Color" },
    forwardColor: { type: ControlType.Color, title: "Forward Color" },
    boolean: { type: ControlType.Boolean, title: "Transition", enabledTitle: "Cube", disabledTitle: "Slide"},
    children: { type: ControlType.Array, title: "Content", propertyControl: { type: ControlType.ComponentInstance, title: "Page" }}
  };

  tapForwards() {
    console.log("Tapped Forwards");
    if (this.state.i < 2) { this.setState({ currentPageIndex: this.state.currentPageIndex + 1, i: this.state.i + 1 }); } 
    else { this.setState({ i: 2 }); } }
  tapBackwards() {
    console.log("Tapped Backwards");
    if (this.state.i > 0) { this.setState({ currentPageIndex: this.state.currentPageIndex - 1, i: this.state.i - 1 }); } 
    else { this.setState({ i: 0 }); }
  }

  render() {
    const { width, height, backgroundColor, backwardColor, forwardColor, ratio, boolean } = this.props;
    const backFrameStyle: React.CSSProperties = {
      width: AreaPagination.defaultProps.width*ratio/100, height: AreaPagination.defaultProps.height,
      top: 0, left: 0,
      position: "absolute",
      backgroundColor: "hsla(0, 0%, 100%, 0)",
    }
    const forwardFrameStyle: React.CSSProperties = {
      width: AreaPagination.defaultProps.width*(1-ratio/100), height: AreaPagination.defaultProps.height,
      top: 0, right: 0,
      position: "absolute",
      backgroundColor: "hsla(0, 0%, 100%, 0)",
    }
    return (
      <div>
        <Page
          {...this.props}
          width={width} height={height}
          gap={0} padding={0}
          background={backgroundColor}
          momentum={false} defaultEffect={boolean ? PageEffect.Cube : PageEffect.None}
          currentPage={this.state.currentPageIndex}
        >
          {this.props.children}
        </Page>
        <Frame style={{...backFrameStyle, width: AreaPagination.defaultProps.width*ratio/100, backgroundColor: backwardColor, }} onClick={this.tapBackwards}></Frame>
        <Frame style={{...forwardFrameStyle, width: AreaPagination.defaultProps.width*(1-ratio/100), backgroundColor: forwardColor}} onClick={this.tapForwards}></Frame>
      </div>
    );
  }
}
