// @framer/bredowmax.pagination/code/ButtonPagination.tsx
import * as React from "react";
import { Frame, Page, PropertyControls, ControlType, PageEffect } from "framer";
import { PageForwards, PageBackwards } from "./canvas";

export interface Props { width: number; height: number; backgroundColor: string; boolean: boolean; } // Define type of property

export class ButtonPagination extends React.Component<Props> {
  constructor(props) {
    super(props);
    this.state = { currentPageIndex: 0, i: 0 };
    this.tapForwards = this.tapForwards.bind(this);
    this.tapBackwards = this.tapBackwards.bind(this);
  }

  static defaultProps = { ...Frame.defaultProps, width: 375, height: 812, backgroundColor: "#000000", boolean: true }; // Set default properties

  static propertyControls: PropertyControls = {
    // Items shown in property panel
    backgroundColor: { type: ControlType.Color, title: "Background" },
    boolean: { type: ControlType.Boolean, title: "Transition", enabledTitle: "Cube", disabledTitle: "Slide" },
    children: { type: ControlType.Array, title: "Content", propertyControl: { type: ControlType.ComponentInstance, title: "Page" } },
  };

  tapForwards() {
    console.log("Tapped Forwards");
    if (this.state.i < 2) { this.setState({ currentPageIndex: this.state.currentPageIndex + 1, i: this.state.i + 1 }); } 
    else { this.setState({ i: 2 }); } }
  tapBackwards() {
    console.log("Tapped Backwards");
    if (this.state.i > 0) { this.setState({ currentPageIndex: this.state.currentPageIndex - 1, i: this.state.i - 1 }); } 
    else { this.setState({ i: 0 }); }
  }

  render() {
    const { width, height, backgroundColor, boolean } = this.props;
    const backFrameStyle: React.CSSProperties = {
      width: 44, height: 44,
      bottom: 20, left: 20,
      position: "absolute",
      backgroundColor: "hsla(0, 0%, 100%, 0)",
    }
    const pageBackward: React.CSSProperties = {
      width: 44, height: 44,
      bottom: 20, left: -30,
      background: "#transparent",
      position: "absolute",
    };
    const forwardFrameStyle: React.CSSProperties = {
      width: 44, height: 44,
      bottom: 20, right: 20,
      position: "absolute",
      backgroundColor: "hsla(0, 0%, 100%, 0)",
    }
    const pageForward: React.CSSProperties = {
      width: 44, height: 44,
      bottom: 20, right: 20,
      position: "absolute",
      background: "#transparent",
    };
    return (
      <div>
        <Page
          {...this.props}
          width={width} height={height}
          gap={0} padding={0}
          background={backgroundColor}
          momentum={false} defaultEffect={boolean ? PageEffect.Cube : PageEffect.None}
          currentPage={this.state.currentPageIndex}
        >
          {this.props.children}
        </Page>
        <Frame style={backFrameStyle} onClick={this.tapBackwards}><PageBackwards style={pageBackward} /></Frame>
        <Frame style={forwardFrameStyle} onClick={this.tapForwards}><PageForwards style={pageForward} /></Frame>
      </div>
    );
  }
}
