// @framer/bredowmax.pagination/code/DragPagination.tsx
import * as React from "react";
import { Frame, Page, PropertyControls, ControlType, PageEffect } from "framer";

export interface Props { width: number; height: number; backgroundColor: string; boolean: boolean; } // Define type of property

export class CubePageNoControls extends React.Component<Props> {

  static defaultProps = { ...Frame.defaultProps, width: 375, height: 812, backgroundColor: "#000000", boolean: true }; // Set default properties

  static propertyControls: PropertyControls = { // Items shown in property panel
    backgroundColor: { type: ControlType.Color, title: "Background" },
    children: { type: ControlType.Array, title: "Content", propertyControl: { type: ControlType.ComponentInstance, title: "Page" }},
    boolean: { type: ControlType.Boolean, title: "Transition", enabledTitle: "Cube", disabledTitle: "Slide", },
  };

  render() {
    const { width, height, backgroundColor, boolean } = this.props;
    
    return (
      <div>
        <Page
          {...this.props}
          width={width} height={height}
          gap={0} padding={0}
          background={backgroundColor}
          momentum={false} defaultEffect={boolean ? PageEffect.Cube : PageEffect.None}
        >
          {this.props.children}
        </Page>
      </div>
    );
  }
}
