// @framer/ismael.photo-filters/code/Filters.tsx
import * as React from "react"
import { Frame, addPropertyControls, ControlType } from "framer"
import "./cssgram.css"

export function Filters({ width, height, resize, image, filter }) {
    let objectFit

    switch (resize) {
        case "fit":
            objectFit = "contain"
            break
        case "fill":
            objectFit = "cover"
            break
        case "stretch":
        default:
            objectFit = "none"
            break
    }

    const styles = {
        wrap: {
            height,
            width,
            overflow: "hidden",
        },
        img: {
            height,
            width,
            display: "block",
            objectPosition: "center",
            objectFit,
        },
    }

    if (image) {
        return (
            <div style={styles.wrap}>
                <div className={filter}>
                    <img src={image} style={styles.img} />
                </div>
            </div>
        )
    } else {
        return (
            <Frame
                size={"100%"}
                style={{
                    justifyContent: "center",
                    display: "flex",
                    flexDirection: "column",
                    alignItems: "center",
                    color: "#8855FF",
                    background: "rgba(136, 85, 255, 0.1)",
                    overflow: "hidden",
                    fontSize: 13,
                    fontWeight: 600,
                }}
            >
                <svg
                    xmlns="http://www.w3.org/2000/svg"
                    width="32"
                    height="32"
                    viewBox="0 0 24 24"
                    fill="none"
                    stroke="#85f"
                    stroke-width="2"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                >
                    <rect x="3" y="3" width="18" height="18" rx="2" ry="2" />
                    <circle cx="8.5" cy="8.5" r="1.5" />
                    <polyline points="21 15 16 10 5 21" />
                </svg>
                <p style={{ marginTop: 5, marginBottom: 0 }}>Choose an image</p>
            </Frame>
        )
    }
}

Filters.defaultProps = {
    image: "",
    filter: "no_filter",
    resize: "fill",
}

addPropertyControls(Filters, {
    image: { type: ControlType.Image, title: "Image" },
    resize: {
        type: ControlType.Enum,
        optionTitles: ["Fill", "Fit", "Stretch"],
        options: ["fill", "fit", "stretch"],
        title: "Resize",
        hidden(props) {
            return props.image === ""
        },
    },
    filter: {
        type: ControlType.Enum,
        options: [
            "no_filter",
            "_1977",
            "aden",
            "brannan",
            "brooklyn",
            "clarendon",
            "earlybird",
            "gingham",
            "hudson",
            "inkwell",
            "kelvin",
            "lark",
            "lo-Fi",
            "naven",
            "nayfair",
            "noon",
            "nashville",
            "perpetua",
            "reyes",
            "rise",
            "slumber",
            "stinson",
            "toaster",
            "valencia",
            "walden",
            "willow",
            "xpro2",
        ],
        optionTitles: [
            "No filter",
            "1977",
            "Aden",
            "Brannan",
            "Brooklyn",
            "Clarendon",
            "Earlybird",
            "Gingham",
            "Hudson",
            "Inkwell",
            "Kelvin",
            "Lark",
            "Lo-Fi",
            "Maven",
            "Mayfair",
            "Moon",
            "Nashville",
            "Perpetua",
            "Reyes",
            "Rise",
            "Slumber",
            "Stinson",
            "Toaster",
            "Valencia",
            "Walden",
            "Willow",
            "X-pro II",
        ],
        title: "Filter",
        hidden(props) {
            return props.image === ""
        },
    },
})
