// @framer/fernandaruiz.visa-credit-card/code/creditcard.tsx
import * as React from "react";
import { PropertyControls, ControlType, Layer } from "framer";

// Define type of property
interface Props {
    firstDigits: string;
    secondDigits: string;
    thirdDigits: string;
    fourthDigits: string;
    expDate: string;
    name: string;
    cardCorners: number;
    topGradient: string;
    bottomGradient: string;
}

const visaSVG: string = `
    <svg enable-background="new 0 0 50 17.4" version="1.1" viewBox="0 0 50 17.4">
        <path d="m24.3 1.2l-2.4 15h-3.9l2.4-15h3.9zm-9.7 0l-3.7 10.3-0.4-2.2-1.4-6.8s-0.2-1.4-1.9-1.4h-6.1l-0.1 0.3h0.1 0.1 0.1 0.1 0.1c0.1 0 0.1 0 0.2 0.1h0.1 0.2 0.1c0.1 0 0.1 0 0.2 0.1h0.1c0.1 0 0.1 0 0.2 0.1h0.1c0 0.3 0.2 0.4 0.4 0.4 0 0 0.1 0 0.1 0.1 0.1 0 0.2 0.1 0.2 0.1 0.1 0 0.1 0.1 0.2 0.1s0.2 0.1 0.2 0.1c0.1 0 0.2 0 0.2 0.1 0.1 0 0.2 0.1 0.3 0.1s0.1 0.1 0.2 0.1c0.1 0.1 0.3 0.2 0.4 0.2l0.2 0.2 3.4 13.1h4.1l6.2-15h-4.2zm34.4 15h-3.6l-0.5-2.2h-5l-0.8 2.2h-4.1l5.8-13.9s0.4-1.1 1.8-1.1h3.1l3.3 15zm-4.7-5.3l-1.2-5.6-2.1 5.6h3.3zm-14.4-5.5c0-1.7 3.8-1.5 5.5-0.6l0.6-3.2s-1.7-0.7-3.5-0.7c-1.9 0-6.6 0.9-6.6 5 0 3.9 5.4 3.9 5.4 6s-4.9 1.7-6.5 0.4l-0.6 3.4s1.7 0.9 4.4 0.9 6.7-1.4 6.7-5.2c0-4-5.4-4.3-5.4-6z" fill="#fff"/>
        <rect x="-38.1" y="-13.3" width="141.7" height="141.7" fill="none"/>
    </svg>`
;

export class creditcard extends React.Component<Props> {

    // Set default properties
    static defaultProps = {
        width: 245,
        height: 155,
        firstDigits: "0000",
        secondDigits: "1111",
        thirdDigits: "2222",
        fourthDigits: "3333",
        expDate: "01/19",
        name: "Maria Fernanda Ruiz",
        cardCorners: 10,
        topGradient: "#030165",
        bottomGradient: "#2E32B8",
    }

    // Items shown in property panel
    static propertyControls: PropertyControls = {
        firstDigits: {type: ControlType.String, title: "First 4 Digits"},
        secondDigits: {type: ControlType.String, title: "Second 4 Digits"},
        thirdDigits: {type: ControlType.String, title: "Third 4 Digits"},
        fourthDigits: {type: ControlType.String, title: "Fourth 4 Digits"},
        expDate: { type: ControlType.String, title: "Expiration Date" },
        name: { type: ControlType.String, title: "Card Holder Name" },
        cardCorners: {type: ControlType.Number, title: "Border Radius"},
        topGradient: {type: ControlType.Color, title: "Top Color"},
        bottomGradient: {type: ControlType.Color, title: "Bottom Color"},
    }

    render() {

        const style: React.CSSProperties = {
            position: "relative",
            width: "100%",
            minWidth: "245px",
            height: "100%",
            minHeight: "155px",
            display: "flex",
            flexDirection: "column",
            justifyContent: "space-between",
            background: `linear-gradient(to bottom, ${this.props.topGradient} 0%,${this.props.bottomGradient} 100%)`,
            overflow: "hidden",
            borderRadius: `${this.props.cardCorners}px`,
            padding: "20px",
        };
        
        const chip: React.CSSProperties = {
            position: "relative",
            width: "20%",
            height: "25%",
            background: "linear-gradient(to bottom, #eeeeee 20%,#999 100%)",
            borderRadius: "7px",
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            flexDirection: "column",
            boxShadow: "0px 0px 5px rgba(0,0,0,.25)",
        }
        
        const digits: React.CSSProperties = {
            width: "100%",
            display: "flex",
            alignItems: "center",
            justifyContent: "space-between",
        }
        
        const expiration: React.CSSProperties = {
            display: "flex",
            alignItems: "flex-end",
            justifyContent: "space-between",
        }

        const bottom: React.CSSProperties = {
            width: "100%",
            display: "flex",
            alignItems: "flex-end",
            justifyContent: "space-between",
        }

        const chipLines: React.CSSProperties = {
            width: "100%",
            height: ".75px",
            backgroundColor: "#999",
            marginTop: "7%",
            marginBottom: "7%",
        }

        const chipMiddle: React.CSSProperties = {
            width: "35%",
            height: "50%",
            borderRadius: "50%",
            background: "linear-gradient(to bottom, #eeeeee 20%,#999 100%)",
            border: "solid .75px #999",
            position: "absolute",
        }

        const glow: React.CSSProperties = {
            width: "100%",
            height: "200%",
            position: "absolute",
            top: "-60%",
            left: "45%",
            transform: "rotate(20deg)",
            borderRadius: "50%",
            background: "linear-gradient(to bottom, rgba(255,255,255,.3) 0%, rgba(255,255,255,0) 80%)",
        }

        const glow2: React.CSSProperties = {
            width: "100%",
            height: "200%",
            position: "absolute",
            top: "-50%",
            left: "50%",
            transform: "rotate(35deg)",
            borderRadius: "50%",
            background: "linear-gradient(to bottom, rgba(255,255,255,.3) 0%, rgba(255,255,255,0) 100%)",
        }

        return(
            <div style={style}>
                <div style={glow} />
                <div style={glow2} />
                <div style={chip}>
                    <span style={chipLines}></span>
                    <span style={chipLines}></span>
                    <span style={chipMiddle}></span>
                </div>
                <div style={digits}>
                    <p style={{ fontSize: "15px", color:"#fff",}}>{this.props.firstDigits}</p>
                    <p style={{ fontSize: "15px", color:"#fff",}}>{this.props.secondDigits}</p>
                    <p style={{ fontSize: "15px", color:"#fff",}}>{this.props.thirdDigits}</p>
                    <p style={{ fontSize: "15px", color:"#fff",}}>{this.props.fourthDigits}</p>
                </div>

                <div style={bottom}>
                    <div style={expiration}>
                        <div>
                            <p style={{ fontSize: "6px", marginBottom: 0, color: "#fff", }}>EXP {this.props.expDate}</p>
                            <p style={{ fontSize: "10px", textTransform: "uppercase", margin: 0, color: "#fff", }}>{this.props.name}</p>
                        </div>
                    </div>
                    <div style={{ width: "20%", overflow: "hidden" }} dangerouslySetInnerHTML={{ __html: visaSVG }} />
                </div>
            </div>
        );
    }
}
