// @framer/bredowmax.progressive-web-app/code/PWA.tsx
import * as React from 'react'
import { ControlType, PropertyControls } from 'framer'

type Props = {
  statusBar: string;
  appTitle: string;
  appIcon: string;
  splashScreen: string;
}

const setTint = value => {
  let statusBar = document.querySelector(
    '[name=apple-mobile-web-app-status-bar-style]'
  )
  if (!statusBar) {
    statusBar = document.createElement('meta')
    statusBar.setAttribute('name', 'apple-mobile-web-app-status-bar-style')
    document.querySelector('head').appendChild(statusBar)
  }
  statusBar.setAttribute('content', value)
}

const setAppTitle = value => {
  let appTitle = document.querySelector(
    '[name=apple-mobile-web-app-title]'
  )
  if (!appTitle) {
    appTitle = document.createElement('meta')
    appTitle.setAttribute('name', 'apple-mobile-web-app-title')
    document.querySelector('head').appendChild(appTitle)
  }
  appTitle.setAttribute('content', value)
}

const setAppIcon = value => {
  let appIcon = document.querySelector(
    '[rel=apple-touch-icon]'
  )
  if (!appIcon) {
    appIcon = document.createElement('link')
    appIcon.setAttribute('rel', 'apple-touch-icon')
    document.querySelector('head').appendChild(appIcon)
  }
  appIcon.setAttribute('href', value)
}

const setSplashScreen = value => {
  let splashScreen = document.querySelector(
    '[rel=apple-touch-startup-image]'
  )
  if (!splashScreen) {
    splashScreen = document.createElement('link')
    splashScreen.setAttribute('rel', 'apple-touch-startup-image')
    document.querySelector('head').appendChild(splashScreen)
  }
  splashScreen.setAttribute('media', "(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)")
  splashScreen.setAttribute('href', value)
}

export class PWA extends React.Component<Props> {
  static defaultProps: Props = {
    statusBar: 'default',
    appTitle: 'Framer X',
    appIcon: 'images/framerApp.png',
    splashScreen: 'images/splashScreen.png',
  }
  
  static propertyControls: PropertyControls = { // Items shown in property panel
    statusBar: {
      type: ControlType.Enum,
      options: ['default', 'black', 'black-translucent'],
      title: "StatusBar",
    },
    appTitle: {
      type: ControlType.String,
      placeholder: "Framer X",
      title: "App Title",
    },
    appIcon: {
      type: ControlType.Image,
      title: 'App Icon',
    },
    splashScreen: {
      type: ControlType.Image,
      title: 'Loading Screen',
    },
  }

  render(){
    setTint(this.props.statusBar)
    setAppTitle(this.props.appTitle)
    setAppIcon(this.props.appIcon)
    setSplashScreen(this.props.splashScreen)
    return <div>Statusbar: {this.props.statusBar} <br/> App Title: {this.props.appTitle}</div>
  }
  
}
